src/main/scala/io/iohk/ethereum/ledger/InMemoryWorldStateProxy.scala

```diff
@@ -0,0 +1,267 @@
+
+    new InMemoryWorldStateProxy(stateStorageProxy, accountsStateTrieProxy, Map.empty, evmCodeStorageProxies, Map.empty, getBlockHashByNumber)
+  }
+
+  def persistIfHashMatches(stateRootHash: ByteString, worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+    val commited = commitState(worldState)
+    if(commited.stateRootHash == stateRootHash) persistStorages(commited)
+    else worldState
+  }
+
+  /**
+    * Updates state trie with current changes but does not persist them into the storages. To do so it:
+    *   - Commits code (to get account's code hashes)
+    *   - Commits constract storages (to get account's contract storage root)
+    *   - Updates state tree
+    *
+    * @param worldState Proxy to commit
+    * @return Updated world
+    */
+  private def commitState(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+    def commitCode(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+      worldState.accountCodes.foldLeft(worldState) {
+        case (updatedWorldState, (address, code)) =>
+          val codeHash = kec256(code)
+          updatedWorldState.copyWith(
+            accountsStateTrie = updatedWorldState.accountsStateTrie +
+              (address.bytes -> updatedWorldState.getGuaranteedAccount(address).copy(codeHash = codeHash)),
+            evmCodeStorage = updatedWorldState.evmCodeStorage + (codeHash -> code),
+            accountCodes = Map.empty
+          )
+      }
+    }
+
+    def commitContractStorage(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy =
+      worldState.contractStorages.foldLeft(worldState) { case (updatedWorldState, (address, storageTrie)) =>
+        val persistedStorage = storageTrie.persist()
+        val newStorageRootHash = persistedStorage.inner.wrapped.getRootHash
+
+        updatedWorldState.copyWith(
+          contractStorages = updatedWorldState.contractStorages + (address -> persistedStorage),
+          accountsStateTrie = updatedWorldState.accountsStateTrie +
+            (address.bytes -> updatedWorldState.getGuaranteedAccount(address).copy(storageRoot = ByteString(newStorageRootHash)))
+        )
+      }
+
+    def commitAccountsStateTrie(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy =
+      worldState.copyWith(accountsStateTrie = worldState.accountsStateTrie.persist())
+
+    (commitCode _ andThen commitContractStorage andThen commitAccountsStateTrie) (worldState)
+  }
+
+  /**
+    * Persists the changes into the storages. This operation will actually write on disk and cannot be undone.
+    *
+    * @param worldState Proxy to persist
+    * @return Updated world
+    */
+  private def persistStorages(worldState: InMemoryWorldStateProxy): InMemoryWorldStateProxy = {
+    worldState.contractStorages.values.foreach(cs => cs.inner.wrapped.nodeStorage.persist())
+    worldState.accountsStateTrie.inner.wrapped.nodeStorage.persist()
```

Comment:
looks like similar case as in 
https://github.com/input-output-hk/etc-client/pull/72/files#diff-7632a397485d58b59d376a5eb28a40c8R76
nesting objects and then manually accessing the most inner box